compiler/transforms/set_memory_space.py

```diff
@@ -0,0 +1,226 @@
+from xdsl.dialects import func, builtin, memref, linalg
+from xdsl.ir import MLContext
+from xdsl.ir.core import SSAValue
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    RewritePattern,
+    PatternRewriter,
+    op_type_rewrite_pattern,
+    PatternRewriteWalker,
+)
+
+
+class InitFuncMemorySpace(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter):
+        """Add a default (0 : i32) memory space to memrefs used in the function
+        that do not have a memory space specified yet"""
+
+        # Function must be public
+        if op.sym_visibility.data != "public":
+            return
+
+        # Function must have memref arguments with an undefined memory space
+        if not any(
+            [
+                isinstance(x, memref.MemRefType)
+                and isinstance(x.memory_space, builtin.NoneAttr)
+                for x in [*op.function_type.inputs, *op.function_type.outputs]
+            ]
+        ):
+            return
+
+        # Mapping function to assign default memory space (0 : i32)
+        def change_to_memory_space(t):
+            if isinstance(t, memref.MemRefType):
+                if isinstance(t.memory_space, builtin.NoneAttr):
+                    return memref.MemRefType.from_element_type_and_shape(
+                        t.element_type,
+                        t.get_shape(),
+                        t.layout,
+                        builtin.IntegerAttr(0, builtin.i32),
+                    )
+            return t
+
+        # Define new function type with updated inputs and outputs
+        # mapped to a default memory space
+        new_function_type = builtin.FunctionType.from_lists(
+            map(change_to_memory_space, op.function_type.inputs),
+            map(change_to_memory_space, op.function_type.outputs),
+        )
+
+        # Change region of function to use new argument types
+        for arg in op.args:
+            arg.type = change_to_memory_space(arg.type)
+
+        # Define new function op with new type and copy region contents
+        new_op = func.FuncOp(
+            op.sym_name.data,
+            new_function_type,
+            region=rewriter.move_region_contents_to_new_regions(op.regions[0]),
+            visibility=op.sym_visibility,
+        )
+
+        # Replace function op
+        rewriter.replace_matched_op(new_op)
+
+
+class InitLinalgMemorySpace(RewritePattern):
```

Comment:
Why do we need to insert a `memory_space_cast` (runtime cast) if we change all `memory_space_casts` in the following pass to be allocs anyway?
I feel like there is only use for this pass if afterwards some `memory_space_cast` will remain, which is not the case I believe?
Is there any `memory_space_cast` that eventually will be turned into risc-v code?

Reply:
Yes! Good question!

I want to stay flexible with memory casts, consider the following scenario:
There is a global address space from address 0 to FF. There are three memory spaces shown in the diagram: 5 - 6 - 8. I want them to be able to overlap each other (In snitch this case may exist for the DMA core which can access L1+L3 while an accelerator only has access to L3)

Now consider the data p1, present in memory space 8 (p1, 8)
If we want to cast it to memory space 6, this is not possible. Therefore, we must allocate + copy. However, in memory space 5 this is not necessary as this memory space already has access to p1. 

This second use case is not yet implemented, but the first conversion to memory_space_casts and only later realizing them allows for this flexibility

```
             ▲
             │xFF
       ┌──   │
       │     │
       │     │
       │8    │
┌──    │     │
│      │     ├──────  p1
│      └──   │
│            │
│5     ┌──   │
│      │     │
│      │     │
│      │     │
│      │     │
└──    │6    │
       │     │
       │     │
       │     │
       └──   │
             │
             │
             │
             │ x00
```

Reply:
I think we can technically still leave them out, but okay to leave them here!